File: 2021/24.py
import z3
import sys

def transpile(program):
    solver = z3.Optimize()

    inp_counter = 0
    inputs = []
    for i in range(14):
        var = z3.BitVec(f"input_{i}", 64)
        solver.add(var > 0)
        solver.add(var < 10)
        inputs += [var]

    model_number = z3.BitVec("model_number", 64)
    expr = inputs[0]
    for inp in inputs[1:]:
        expr = expr * 10 + inp
    solver.add(model_number == expr)

    fresh_counter = 0
    def fresh():
        nonlocal fresh_counter
        var = z3.BitVec(f"fresh_{fresh_counter}", 64)
        fresh_counter += 1
        return var

    registers = {"w": 0, "x": 0, "y": 0, "z": 0}
    def term(t):
        nonlocal registers
        if t.isalpha():
            return registers[t]
        else:
            return int(t)

    for line in program.splitlines():
        words = line.split()
        if words[0] == "inp":
            registers[words[1]] = inputs[inp_counter]
            inp_counter += 1
        elif words[0] == "add":
            lhs = fresh()
            solver.add(lhs == term(words[1]) + term(words[2]))
            registers[words[1]] = lhs
        elif words[0] == "mul":
            lhs = fresh()
            solver.add(lhs == term(words[1]) * term(words[2]))
            registers[words[1]] = lhs
        elif words[0] == "div":
            lhs = fresh()
            solver.add(term(words[2]) != 0)
            solver.add(lhs == term(words[1]) / term(words[2]))
            registers[words[1]] = lhs
        elif words[0] == "mod":
            lhs = fresh()
            solver.add(term(words[2]) != 0)
            solver.add(lhs == term(words[1]) % term(words[2]))
            registers[words[1]] = lhs
        elif words[0] == "eql":
            lhs = fresh()
            solver.add(lhs == z3.If(term(words[1]) == term(words[2]), z3.BitVecVal(1, 64), z3.BitVecVal(0, 64)))
            registers[words[1]] = lhs
        else:
            raise Exception("boooo")

    solver.add(registers["z"] == 0)
    return solver, model_number

input = sys.stdin.read()

solver, model_number = transpile(input)
solver.maximize(model_number)
if solver.check() != z3.sat:
    raise Exception("booooo")
print(solver.model()[model_number])

solver, model_number = transpile(input)
solver.minimize(model_number)
if solver.check() != z3.sat:
    raise Exception("booooooo")
print(solver.model()[model_number])

File: 2021/24.hs
{-# LANGUAGE DataKinds #-}
import           AdventOfCode.Main
import qualified AdventOfCode.NanoParser as NP
import qualified AdventOfCode.Z3         as Z3
import qualified Data.Map as M
import           Control.Applicative     (many, (<|>))

type Program = [Instruction]

data Register = W | X | Y | Z deriving (Bounded, Enum, Eq, Ord, Show)

data Operand = Literal Int | Register Register deriving (Show)

data Instruction
    = Inp Register
    | Add Register Operand
    | Mul Register Operand
    | Div Register Operand
    | Mod Register Operand
    | Eql Register Operand
    deriving (Show)

parseInstruction :: NP.Parser Char Instruction
parseInstruction =
    (Inp <$> (tok (NP.string "inp") *> tok register)) <|>
    (Add <$> (tok (NP.string "add") *> tok register) <*> tok operand) <|>
    (Mul <$> (tok (NP.string "mul") *> tok register) <*> tok operand) <|>
    (Div <$> (tok (NP.string "div") *> tok register) <*> tok operand) <|>
    (Mod <$> (tok (NP.string "mod") *> tok register) <*> tok operand) <|>
    (Eql <$> (tok (NP.string "eql") *> tok register) <*> tok operand)
  where
    tok p    = p <* NP.spaces
    operand  = (Literal <$> NP.signedDecimal) <|> (Register <$> register)
    register =
        (W <$ NP.char 'w') <|>
        (X <$ NP.char 'x') <|>
        (Y <$ NP.char 'y') <|>
        (Z <$ NP.char 'z')

parseProgram :: NP.Parser Char Program
parseProgram = many parseInstruction

toZ3 :: Bool -> Program -> Z3.Program
toZ3 part1 program = mconcat $
    [ Z3.declareConst i <>
        Z3.assert (Z3.bvsgt (Z3.var i) (Z3.int2bv (Z3.int 0))) <>
        Z3.assert (Z3.bvslt (Z3.var i) (Z3.int2bv (Z3.int 10)))
    | i <- inputs] <>
    [Z3.declareConst modelNumber] <>
    [Z3.assert $ Z3.var modelNumber Z3.== foldl1
        (\acc i -> Z3.bvadd i (Z3.bvmul acc (Z3.int2bv (Z3.int 10))))
        (map Z3.var inputs)] <>
    [go 0 0 initRegisters program] <>
    [(if part1 then Z3.maximize else Z3.minimize) (Z3.var modelNumber)] <>
    [Z3.checkSat] <>
    [Z3.eval (Z3.bv2int (Z3.var modelNumber))]
  where
    inputs :: [Z3.Var ('Z3.BitVecSort 64)]
    inputs = [Z3.mkVar $ "input" ++ show i | i <- [0 :: Int .. 13]]

    modelNumber :: Z3.Var ('Z3.BitVecSort 64)
    modelNumber = Z3.mkVar "modelnumber"

    initRegisters :: M.Map Register (Z3.Expr ('Z3.BitVecSort 64))
    initRegisters = M.fromList $ do
        reg <- [minBound .. maxBound]
        pure (reg, Z3.int2bv (Z3.int 0))

    operand
        :: M.Map Register (Z3.Expr ('Z3.BitVecSort 64))
        -> Operand
        -> Z3.Expr ('Z3.BitVecSort 64)
    operand regs (Register x) = regs M.! x
    operand regs (Literal x)  = Z3.int2bv (Z3.int x)

    go :: Int -> Int -> M.Map Register (Z3.Expr ('Z3.BitVecSort 64)) -> Program -> Z3.Program
    go ic fc regs instructions = case instructions of
        [] ->
            Z3.assert ((regs M.! Z) Z3.== Z3.int2bv (Z3.int 0)) <> []
        Inp reg : instrs ->
            go (ic + 1) fc (M.insert reg (Z3.var $ inputs !! ic) regs) instrs
        Add x y : instrs ->
            let fresh = Z3.mkVar $ "fresh" ++ show fc :: Z3.Var ('Z3.BitVecSort 64) in
            Z3.declareConst fresh <>
            Z3.assert (Z3.var fresh Z3.== Z3.bvadd (regs M.! x) (operand regs y)) <>
            go ic (fc + 1) (M.insert x (Z3.var fresh) regs) instrs
        Mul x y : instrs ->
            let fresh = Z3.mkVar $ "fresh" ++ show fc :: Z3.Var ('Z3.BitVecSort 64) in
            Z3.declareConst fresh <>
            Z3.assert (Z3.var fresh Z3.== Z3.bvmul (regs M.! x) (operand regs y)) <>
            go ic (fc + 1) (M.insert x (Z3.var fresh) regs) instrs
        Div x y : instrs ->
            let fresh = Z3.mkVar $ "fresh" ++ show fc :: Z3.Var ('Z3.BitVecSort 64) in
            Z3.declareConst fresh <>
            Z3.assert (Z3.not $ operand regs y Z3.== Z3.int2bv (Z3.int 0)) <>
            Z3.assert (Z3.var fresh Z3.== Z3.bvsdiv (regs M.! x) (operand regs y)) <>
            go ic (fc + 1) (M.insert x (Z3.var fresh) regs) instrs
        Mod x y : instrs ->
            let fresh = Z3.mkVar $ "fresh" ++ show fc :: Z3.Var ('Z3.BitVecSort 64) in
            Z3.declareConst fresh <>
            Z3.assert (Z3.not $ operand regs y Z3.== Z3.int2bv (Z3.int 0)) <>
            Z3.assert (Z3.var fresh Z3.== Z3.bvsmod (regs M.! x) (operand regs y)) <>
            go ic (fc + 1) (M.insert x (Z3.var fresh) regs) instrs
        Eql x y : instrs ->
            let fresh = Z3.mkVar $ "fresh" ++ show fc :: Z3.Var ('Z3.BitVecSort 64) in
            Z3.declareConst fresh <>
            Z3.assert (Z3.var fresh Z3.== Z3.ite ((regs M.! x) Z3.== (operand regs y))
                (Z3.int2bv (Z3.int 1)) (Z3.int2bv (Z3.int 0))) <>
            go ic (fc + 1) (M.insert x (Z3.var fresh) regs) instrs

main :: IO ()
main = ioMain $ \str -> do
    program <- either fail pure $ NP.runParser parseProgram str
    pure (Z3.run "2021-24" $ toZ3 True program, Z3.run "2021-24" $ toZ3 False program)
